connector/connector_ldap.go

```diff
@@ -217,11 +250,108 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
 	return c.trustedEmailProvider
 }
 
+// A LDAPPool is a Connection Pool for LDAP connections
+// Initialize exported fields and use Get() to get a connection.
+// Use Put() to put it back into the pool.
+type LDAPPool struct {
+	m              sync.Mutex
+	conns          map[*ldap.Conn]struct{}
+	MaxIdleConn    int
+	PoolCheckTimer time.Duration
+	ServerHost     string
+	ServerPort     uint16
+	UseTLS         bool
+	UseSSL         bool
+	TlsConfig      *tls.Config
+}
+
+func (p *LDAPPool) Get(params ...*ldap.Conn) (*ldap.Conn, error) {
+	var conn *ldap.Conn
+
+	p.m.Lock()
```

Comment:
What's being guarded by the lock and what's intentionally outside the lock isn't obvious from this flow, a lot of conditions, and manual unlocks. Splitting out the functionality of Get() and Get(conn) should help.